llama.cpp

```diff
@@ -0,0 +1,1556 @@
+                }
+                printf("\n");
+            } else {
+                printf("size = %8.3f MB\n", data_u8.size()/1024.0/1024.0);
+                fout.write(reinterpret_cast<char *>(data_u8.data()), data_u8.size());
+                total_size_new += data_u8.size();
+            }
+
+            total_size_org += nelements * sizeof(float);
+        }
+
+        printf("%s: model size  = %8.2f MB\n", __func__, total_size_org/1024.0/1024.0);
+        printf("%s: quant size  = %8.2f MB\n", __func__, total_size_new/1024.0/1024.0);
+
+        {
+            int64_t sum_all = 0;
+            for (int i = 0; i < (int) hist_all.size(); ++i) {
+                sum_all += hist_all[i];
+            }
+
+            printf("%s: hist: ", __func__);
+            for (int i = 0; i < (int) hist_all.size(); ++i) {
+                printf("%5.3f ", hist_all[i] / (float)sum_all);
+            }
+            printf("\n");
+        }
+    }
+
+    finp.close();
+    fout.close();
+
+    return true;
+}
+
+//
+// interface implementation
+//
+
+struct llama_context * llama_init_from_file(
+                             const char * path_model,
+            struct llama_context_params   params) {
+    ggml_time_init();
+
+    llama_context * ctx = new llama_context;
+
+    ctx->rng = std::mt19937(params.seed);
+    ctx->logits_all = params.logits_all;
+
+    ggml_type type_memory = params.f16_kv ? GGML_TYPE_F16 : GGML_TYPE_F32;
+
+    if (!llama_model_load(path_model, *ctx, params.n_ctx, params.n_parts, type_memory, params.vocab_only)) {
+        fprintf(stderr, "%s: failed to load model\n", __func__);
+        delete ctx;
+        return nullptr;
+    }
+
+    return ctx;
+}
+
+void llama_free(struct llama_context * ctx) {
+    ggml_free(ctx->model.ctx);
+
+    delete ctx;
+}
+
+int llama_model_quantize(
+        const char * fname_inp,
+        const char * fname_out,
+               int   itype,
+               int   qk) {
+    if (!llama_model_quantize_internal(fname_inp, fname_out, itype, qk)) {
+        fprintf(stderr, "%s: failed to quantize\n", __func__);
+        return 1;
+    }
+
+    return 0;
+}
+
+int llama_eval(
+        struct llama_context * ctx,
+           const llama_token * tokens,
+                         int   n_tokens,
+                         int   n_past,
+                         int   n_threads) {
+    if (!llama_eval_internal(*ctx, tokens, n_tokens, n_past, n_threads)) {
+        fprintf(stderr, "%s: failed to eval\n", __func__);
+        return 1;
+    }
+
+    return 0;
+}
+
+int llama_tokenize(
+        struct llama_context * ctx,
+                  const char * text,
+                 llama_token * tokens,
+                         int   n_max_tokens,
+                        bool   add_bos) {
+    auto res = llama_tokenize(ctx->vocab, text, add_bos);
+
+    if (n_max_tokens < (int) res.size()) {
+        fprintf(stderr, "%s: too many tokens\n", __func__);
+        return 1;
```

Comment:
Returning 1 here is ambiguous, it is impossible to tell if there was an error or if only 1 token was returned. Consider returning a negative number to indicate failure (maybe the number of tokens negated?).

Reply:
According to the documentation in llama.h it should return -1 on failure. However, may be necessary to rethink or add other API to support cases like the perplexity computation in which a big document has to be tokenized.

Reply:
Good catch. Lets fix the perplexity computation on `master` to avoid blocking the development for too long.
With the new API change that you suggested, we can now query for the number of needed tokens and allocate the necessary memory. Some improvements might also be possible